### PR TITLE
WIP: Register machine `wasmi` execution engine (take 2)

### PR DESCRIPTION
**Pull request by Robbepop**
Closes https://github.com/paritytech/wasmi/issues/361.

Precursor: https://github.com/paritytech/wasmi/pull/367

## ToDo

These items are things we want to do before merging the PR:

- [ ] Simplify call argument encoding for function calls with exactly 1 argument.

For function calls we need to setup the arguments so that they are all stored in contiguous registers. Currently we do this for function calls with exactly 1 argument even though this is not needed. Thus during translation we should implement a check and simplify (and thus optimize) function call argument encoding for those function calls.

- [ ] Properly implement translation of `ConsumeFuel` instructions and their fixed costs.

So far we have concentrated on getting the Wasm to `wasmi` bytecode translation up and running with the most simple setup possible. This means we ignored translation of `ConsumeFuel` instructions and their associated fuel costs so far and are in need of doing this work before we merge the PR. There should not be any difficulties compared to the already existing implementation for the stack-machine engine backend.

- [x] Fix attack vector of `local.set` stack preservation:

Currently for every `local.set` or `local.tee` that we encounter while translating Wasm bytecode to `wasmi` bytecode we iterate all values on the emulated value stack. Usually this isn't a big deal since the stack usually doesn't grow big in practical workloads. However, this can be easily attacked by a Wasm blob that blows up the stack and then performs tons of `local.set` instructions. We can eliminate this attack vector by storing the `local.get` provider indices on the emulated stack on a separate stack and iterate on it instead. While iterating we also remove the `local.get` indices from the stack so that a consecutive `local.set` operation will see an empty stack and thus perform no operations. We simply cache the `local.get` providers this way.

- [ ] Improve `local.set` result replacement optimization when preserving `local.get` at the same time:

When translating `local.set` or `local.tee` we replace the result register of the previous instruction instead of emitting a copy instruction if possible. However, when preserving a `local.get` on the emulation stack we do not perform this optimization. The problem is that the copy instruction required for the preservation is required to take place before the instruction `i` that should have its result register replaced. However, the instruction `i` is already encoded and could in the worst case consist of multiple instruction words which would require to shift already encoded instructions by one index. In theory this could also interfere with pre-calculated branch offset calculations but so far this has not been demonstrated and might not be true.

- [x] Avoid expensive register space defragmentation if possible.

After successful translation of Wasm bytecode the `wasmi` translation performs a final pass over all encoded instructions to defragment the register space. This is needed for registers that have been preserved for `local.set` in certain situations. However, this entire process is very costly and can be avoided entirely or partially. We can avoid it entirely since this only ever needs to be done if there were actual register preservations during the translation procesure. Furthermore we only need to defragment all instructions that have been encoded after encountering the first `local.set` register preservation. This way we can keep the simple loop over all instructions but still avoid most of the unnecessary work and thus speed up the translation performance.

- [ ] Recycle preservation register slots if possible.

Currently, when preserving `local.get x` values on the stack upon a `local.set x` translation, a new register is reserved for the N `local.get x` values that have been found and replaced on the emulated value stack at the time of translation. When this happens multiple times, a new register slot on the preservation stack is registered each time. Right now, we do not track how many of the preserved `local.get x` values have already been used. However, if we would do this we could recycle no longer used preservation register slots instead of allocating a new slot all the time which could lead to fewer registers used especially by larger functions. A major data structure that would allow this efficiently in O(1) is the so-called Stash data structure.

## Plan & Steps

- [x] Define the new register-machine based bytecode for the new `wasmi` executor.
    - [x] Branch instructions
        - [x] Unconditional Branches
        - [x] Conditional Branches
        - [x] Branch Table
    - [x] Trap Instructions, e.g. Wasm `unreachable` instruction
    - [x] Wasm `select` and `select <ty>` Instructions
    - [x] Return Instructions
        - [x] Unconditional Returns
        - [x] Conditional Returns
    - [x] Call Instructions
        - [x] Nested Calls
            - [x] Internal Function Calls
            - [x] Imported Function Calls
            - [x] Indirect Function Calls
        - [x] Tail Calls
            - [x] Internal Function Calls
            - [x] Imported Function Calls
            - [x] Indirect Function Calls
    - [x] Comparison Instructions (or equivalents)
        - [x] `i32.{eq, eqz, ne, lt_{s|u}, le_{s|u}, gt_{s|u}, ge_{s|u}}` instructions
        - [x] `i64.{eq, eqz, ne, lt_{s|u}, le_{s|u}, gt_{s|u}, ge_{s|u}}` instructions
        - [x] `f32.{eq, ne, lt, le, gt, ge}` instructions
        - [x] `f64.{eq, ne, lt, le, gt, ge}` instructions
    - [x] Memory `load` Instructions (or equivalents)
        - [x] `i32.load` and `i32.loadN_{s|u}` instructions
        - [x] `i64.load` and `i64.loadN_{s|u}` instructions
        - [x] `f32.load` instruction
        - [x] `f64.load` instruction
    - [x] Memory `store` instructions (or equivalents)
        - [x] `i32.store` and `i32.storeN` instructions
        - [x] `i64.store` and `i64.storeN` instructions
        - [x] `f32.store` instruction
        - [x] `f64.store` instruction
    - [x] Compute Instructions (or equivalents)
        - [x] `i32` compute instructions, e.g. `i32.popcnt`, `i32.add`, `i32.rotl` etc..
        - [x] `i64` compute instructions, e.g. `i64.popcnt`, `i64.add`, `i32.rotl` etc..
        - [x] `f32` compute instructions, e.g. `f32.sqrt`, `f32.add`, etc..
        - [x] `f64` compute instructions, e.g. `f32.sqrt`, `f32.add`, etc..
    - [x] Conversion Instructions
        - [x] MVP conversion instructions
        - [x] `sign-extension` proposal instructions
        - [x] `non-trapping float-to-int conversion` proposal instructions
    - [x] Global Variable Instructions (or equivalents)
        - [x] `global.get`
        - [x] `global.set` (and immediate versions)
    - [x] Wasm `table` instructions
        - [x] `table.size` instruction
        - [x] `table.grow` instruction
        - [x] `table.get` instruction
        - [x] `table.set` instruction
        - [x] `table.fill` instruction
        - [x] `table.copy` instruction
        - [x] `table.init` instruction
    - [x] Wasm `memory` instructions
        - [x] `memory.size` instruction
        - [x] `memory.grow` instruction
        - [x] `memory.fill` instruction
        - [x] `memory.copy` instruction
        - [x] `memory.init` instruction
- [x] Translation of Wasm bytecode to register-machine based `wasmi` bytecode
    - **Note:** This includes translation as well as proper unit tests coverage for all `wasmi` instructions.
    - [x] Control Flow
        - [x] `block` control flow
        - [x] `loop` control flow
        - [x] `if` control flow
    - [x] Branches
        - [x] Unconditional Branches
        - [x] Conditional Branches
        - [x] Branch Table: `br_table`
    - [x] Returns
    - [x] Call Instructions
        - [x] Nested Calls
            - [x] Internal Function Calls
            - [x] Imported Function Calls
            - [x] Indirect Function Calls
        - [x] Tail Calls
            - [x] Internal Function Calls
            - [x] Imported Function Calls
            - [x] Indirect Function Calls
    - [x] Select Instruction
        - [x] Untyped `select` (from Wasm MVP)
        - [x] Typed `select (result <ty>)` (from `reference-types` proposal)
    - [x] Wasm `drop` instruction
    - [x] Wasm `unreachable` instruction
    - [x] Wasm `local.set` and `local.tee`
    - [x] Comparison Instructions (or equivalents)
        - [x] `i32.{eq, ne, eqz}` instructions
        - [x] `i64.{eq, ne, eqz}` instructions
        - [x] `f32.{eq, ne}` instructions
        - [x] `f64.{eq, ne}` instructions
        - [x] `i32.{lt_s, lt_u, le_s, le_u, gt_s, gt_u, ge_s, ge_u}` instructions
        - [x] `i64.{lt_s, lt_u, le_s, le_u, gt_s, gt_u, ge_s, ge_u}` instructions
        - [x] `f32.{lt, le, gt, ge}` instructions
        - [x] `f64.{lt, le, gt, ge}` instructions
    - [x] Memory `load` Instructions (or equivalents)
        - [x] `i32.load` and `i32.loadN_{s|u}` instructions
        - [x] `i64.load` and `i64.loadN_{s|u}` instructions
        - [x] `f32.load` instruction
        - [x] `f64.load` instruction
    - [x] Memory `store` instructions (or equivalents)
        - [x] `i32.store` and `i32.storeN` instructions
        - [x] `i64.store` and `i64.storeN` instructions
        - [x] `f32.store` instruction
        - [x] `f64.store` instruction
    - [x] Compute instructions
        - [x] Unary instructions
            - [x] `i32.{clz, ctz, popcnt}`
            - [x] `i64.{clz, ctz, popcnt}`
            - [x] `f32.{abs, neg, ceil, floor, trunc, nearest, sqrt}`
            - [x] `f64.{abs, neg, ceil, floor, trunc, nearest, sqrt}`
        - [x] Commutative instructions
            - [x] `i32.{add, mul, and, or, xor}`
            - [x] `i64.{add, mul, and, or, xor}`
            - [x] `f32.{add, mul, min, max}`
            - [x] `f64.{add, mul, min, max}`
        - [x] Non-commutative instructions
            - [x] `i32.sub`
            - [x] `i64.sub`
            - [x] `f32.{sub, div, copysign}`
            - [x] `f64.{sub, div, copysign}`
        - [x] Shift instructions
            - [x] `i32.{shl, shr_s, shr_u, rotl, rotr}`
            - [x] `i64.{shl, shr_s, shr_u, rotl, rotr}`
        - [x] Divide and remainder instructions
            - [x] `i32.{div_u, div_s, rem_u, rem_s}`
            - [x] `i64.{div_u, div_s, rem_u, rem_s}`
    - [x] Conversion Instructions
        - [x] MVP conversion instructions
        - [x] `sign-extension` proposal instructions
        - [x] `non-trapping float-to-int conversion` proposal instructions
    - [x] Global Variable Instructions (or equivalents)
        - [x] `global.get` instruction
        - [x] `global.set` instruction
    - [x] Wasm `reftype` instructions
        - [x] `ref.null`
        - [x] `ref.is_null`
        - [x] `ref.func`
    - [x] Wasm `table` instructions
        - [x] `table.size` instruction
        - [x] `table.grow` instruction
        - [x] `table.get` instruction
        - [x] `table.set` instruction
        - [x] `table.fill` instruction
        - [x] `table.copy` instruction
        - [x] `table.init` instruction
        - [x] `elem.drop` instruction
    - [x] Wasm `memory` instructions
        - [x] `memory.size` instruction
        - [x] `memory.grow` instruction
        - [x] `memory.fill` instruction
        - [x] `memory.copy` instruction
        - [x] `memory.init` instruction
        - [x] `data.drop` instruction
- [x] Execution of `wasmi` register-machine bytecode
    - [x] Trap instruction
    - [x] Consume fuel instruction
    - [x] Branch instructions
    - [x] Return instructions
    - [x] Copy instructions
    - [x] Select instructions
    - [x] Wasm `ref.func` instruction
    - [x] Call instructions
        - [x] Internal calls
            - [x] Nested
            - [x] Tail
        - [x] Imported calls
            - [x] Nested
            - [x] Tail
        - [x] Indirect calls
            - [x] Nested
            - [x] Tail
        - [x] Host calls
    - [x] Wasm `load` instructions
    - [x] Wasm `store` instructions
    - [x] Wasm `table` instructions
        - [x] `table.get` instructions
        - [x] `table.set` instructions
        - [x] `table.size` instruction
        - [x] `table.copy` instruction
        - [x] `table.init` instruction
        - [x] `table.fill` instruction
        - [x] `table.grow` instruction
        - [x] `elem.drop` instruction
    - [x] Wasm `memory` instructions
        - [x] `memory.size` instruction
        - [x] `memory.copy` instruction
        - [x] `memory.init` instruction
        - [x] `memory.fill` instruction
        - [x] `memory.grow` instruction
        - [x] `data.drop` instruction
    - [x] Wasm `global` instructions
        - [x] `global.get` instructions
        - [x] `global.set` instructions
    - [x] Unary instructions
        - [x] `i32` instructions
        - [x] `i64` instructions
        - [x] `f32` instructions
        - [x] `f64` instructions
    - [x] Comparison instructions
    - [x] Shift & rotate instructions
    - [x] Binary instructions
        - [x] `i32` instructions
        - [x] `i64` instructions
        - [x] `f32` instructions
        - [x] `f64` instructions
    - [x] Conversion instructions

## Ideas

The following list contains ideas that spun up and might be iterated here for experimentation purposes.

- [ ] Wasm `br 0` in Wasm basic `block` control frames should not be translated as a branch and instead as a Wasm `end` of the basic `block` since all code after the `br 0` is unreachable. Fortunately this bytecode sequence seems to not be very common in practical Wasm blobs.
- [ ] Use inline encoding for 64-bit immediate instructions.
    - **Note:** At the moment 64-bit encoded immediates use `ConstRef` in order to store the actual value in a const pool which is external to the bytecode itself. This has several downsides for performance and also for bytecode integrity. Performance is affected since an additional indirect memory fetch is required in order to compute on the constant value. Bytecode integrity is worse since analysing `wasmi` bytecode now also needs to inspect the external const pool. The latter point affects testability of `wasmi` bytecode. A way to improve this situation in both areas is to store the 64-bit constants inline. The problem is that for this 8 bytes are required but instruction words only support up to 6 bytes of parameters per word. Thus there is a need to split up the 64-bit constant into multiple pieces. The experiment [in this GitHub Gist](https://gist.github.com/Robbepop/3173b3f2e3eb4632e1771cbacecec7af) shows that a solution that splits the 64-bit constant value into 3 pieces (2 x 2-byte and 1 x 4-byte pieces) can be done efficiently on x86 and Wasm platforms. However, further benchmark tests are required to proof this.
- [ ] For `{i32, i64}.{div_u, div_s, rem_u, rem_s}` we can apply a bytecode encoding optimization for the cases where the right-hand side divisor is a constant value. During translation we guarantee for all those operations that the right-hand side constant value is non-zero due to the fact that a zero right-hand side value is translated as a `trap` instruction during translation. Therefore we can replace the `Const32` or `Const16` parameter of those instructions with a `NonZero32` or `NonZero16` value and use Rust's built-in `Div<NonZeroU{32,64}> for u{32,64}` and `Rem<NonZeroU{32,64}> for u{32,64}`. Note that those APIs are only available for unsigned integers.
- [ ] Add op-assign instructions.
    - **Note:** The idea is to introduce op-assign instructions such as `i32.add_assign`. The advantage is that we no longer have to store both, `result` and `lhs` fields since they are always the same and always of type `Register`. This also allows to have inline `Const32` or `ConstRef` `rhs` fields and thus we could save some encoding space, too. Another benefit is that the instruction is probably also a bit faster at execution since the compiler has more information about which registers to read to and write from. However, this has to be checked with experiments. A downside of having these op-assign instructions is that they do not play well with `local.set` and `local.tee` optimizations where we replace the `result` register of the previous instruction during translation phase
    - Example instructions for `AddAssign` or `+=` operator:
        - `I32AddAssign(UnaryInstr)`
        - `I32AddAssignImm(UnaryInstrImm32)`: Requires just 1 `Instruction` for encoding.
        - `I64AddAssign(UnaryInstr)`
        - `I64AddAssignImm(UnaryInstrImm)`: Requires just 1 `Instruction` for encoding.
        - `I64AddAssignImm32(UnaryInstrImm32)`: 32-bit small value optimization instead of 16-bit
        - `F32AddAssign(UnaryInstr)`
        - `F32AddAssignImm(UnaryInstrImm32)`: Requires just 1 `Instruction` for encoding.
        - `F64AddAssign(UnaryInstr)`
        - `F64AddAssignImm(UnaryInstrImm)`: Requires just 1 `Instruction` for encoding.
- [ ] Fuse comparison and branch instructions as proposed in [this issue](https://github.com/paritytech/wasmi/issues/712).
    - **Note:** The idea behind this optimization is that conditional branch instructions often follow comparison instructions and thus this sequence can easily be identified and optimized. Furthermore both instructions are very common in execution hot paths, for example in loops and thus optimizing them has greater effects.
- [ ] Introduce `global.op_assign` instructions
    - These special instructions act as an optimization on Wasm instruction sequences such as `global.get g; i32.add; global.set g` which we then could represent using a single `wasmi` instructions such as `global.i32.add g r` or `global.i32.add_imm g c` where `g` represents a global, `r` represents an input register and `c` a constant value. Further research is needed to find out how common these sequences are and if the proposed `global.op_assign` instruction are actually improving execution performance significantly.
    - **Note:** It is likely that global variables are very sparsely used and thus this optimization likely won't improve performance too much while adding tons of new special instructions to `wasmi` bytecode. For example `spidermonkey.wasm` contains exactly a single global variable that is only sparsely used thoughout the Wasm file.
- [ ] Add pointer arithmetic to `load` and `store` instructions.
    - **Note:** Fuse instruction sequences for `load` and `store` instructions that oddly compute a `ptr+offset` or `ptr*scale + offset` or even `ptr shift scale + offset` outside of Wasm's `load` and `store` instructions. Technically we can make `load` and `store` instructions very powerful by including simple pointer arithmetic into these instructions.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
## BENCHMARKS ## 

 <table> <thead> <tr> <th></th><th colspan=3>NATIVE</th><th colspan=3>WASMTIME</th><th></th> </tr> <tr> <th>BENCHMARK</th><th>MASTER</th><th>PR</th><th>DIFF</th><th>MASTER</th><th>PR</th><th>DIFF</th><th>WASMTIME OVERHEAD</th> </tr> </thead> <tbody> <tr><td><tt>execute/</tt><br><tt>bare_call_0</td> <td nowrap> 1.55ms</td> <td nowrap> 1.61ms</td> <td nowrap> :red_circle: 4.28%</td> <td nowrap> 1.04ms</td> <td nowrap> 1.06ms</td> <td nowrap> :green_circle: 1.25%</td> <td nowrap> :green_circle: -35%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_0/typed</td> <td nowrap> 1.18ms</td> <td nowrap> 1.31ms</td> <td nowrap> :red_circle: 10.83%</td> <td nowrap> 678.62µs</td> <td nowrap> 655.56µs</td> <td nowrap> :green_circle: -3.32%</td> <td nowrap> :green_circle: -50%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_1</td> <td nowrap> 1.59ms</td> <td nowrap> 1.65ms</td> <td nowrap> :red_circle: 2.96%</td> <td nowrap> 1.19ms</td> <td nowrap> 1.17ms</td> <td nowrap> :green_circle: -2.59%</td> <td nowrap> :green_circle: -29%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_16</td> <td nowrap> 2.46ms</td> <td nowrap> 2.46ms</td> <td nowrap> :red_circle: 0.09%</td> <td nowrap> 3.23ms</td> <td nowrap> 3.23ms</td> <td nowrap> :white_circle: -0.02%</td> <td nowrap> :green_circle: 31%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_16/typed</td> <td nowrap> 1.52ms</td> <td nowrap> 1.56ms</td> <td nowrap> :red_circle: 2.86%</td> <td nowrap> 1.67ms</td> <td nowrap> 1.65ms</td> <td nowrap> :white_circle: -1.00%</td> <td nowrap> :green_circle: 6%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_1/typed</td> <td nowrap> 1.29ms</td> <td nowrap> 1.29ms</td> <td nowrap> :white_circle: -0.61%</td> <td nowrap> 995.21µs</td> <td nowrap> 976.60µs</td> <td nowrap> :green_circle: -1.86%</td> <td nowrap> :green_circle: -24%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_4</td> <td nowrap> 1.73ms</td> <td nowrap> 1.83ms</td> <td nowrap> :red_circle: 5.13%</td> <td nowrap> 1.57ms</td> <td nowrap> 1.59ms</td> <td nowrap> :white_circle: 1.12%</td> <td nowrap> :green_circle: -13%</td></tr><tr><td><tt>execute/</tt><br><tt>bare_call_4/typed</td> <td nowrap> 1.20ms</td> <td nowrap> 1.43ms</td> <td nowrap> :red_circle: 19.07%</td> <td nowrap> 916.77µs</td> <td nowrap> 906.48µs</td> <td nowrap> :white_circle: -1.18%</td> <td nowrap> :green_circle: -36%</td></tr><tr><td><tt>execute/</tt><br><tt>br_table</td> <td nowrap> 1.34ms</td> <td nowrap> 1.39ms</td> <td nowrap> :red_circle: 3.85%</td> <td nowrap> 1.19ms</td> <td nowrap> 1.19ms</td> <td nowrap> :white_circle: -0.22%</td> <td nowrap> :green_circle: -14%</td></tr><tr><td><tt>execute/</tt><br><tt>count_until</td> <td nowrap> 679.96µs</td> <td nowrap> 714.87µs</td> <td nowrap> :red_circle: 5.91%</td> <td nowrap> 1.28ms</td> <td nowrap> 1.28ms</td> <td nowrap> :white_circle: -0.24%</td> <td nowrap> :yellow_circle: 78%</td></tr><tr><td><tt>execute/</tt><br><tt>factorial_iterative</td> <td nowrap> 318.84µs</td> <td nowrap> 317.72µs</td> <td nowrap> :white_circle: -0.37%</td> <td nowrap> 537.31µs</td> <td nowrap> 536.77µs</td> <td nowrap> :white_circle: -0.11%</td> <td nowrap> :yellow_circle: 69%</td></tr><tr><td><tt>execute/</tt><br><tt>factorial_recursive</td> <td nowrap> 511.19µs</td> <td nowrap> 510.26µs</td> <td nowrap> :white_circle: -0.06%</td> <td nowrap> 668.33µs</td> <td nowrap> 697.35µs</td> <td nowrap> :red_circle: 4.23%</td> <td nowrap> :green_circle: 37%</td></tr><tr><td><tt>execute/</tt><br><tt>fibonacci_iter</td> <td nowrap> 1.40ms</td> <td nowrap> 1.40ms</td> <td nowrap> :white_circle: -0.04%</td> <td nowrap> 2.65ms</td> <td nowrap> 2.65ms</td> <td nowrap> :white_circle: 0.06%</td> <td nowrap> :yellow_circle: 90%</td></tr><tr><td><tt>execute/</tt><br><tt>fibonacci_rec</td> <td nowrap> 4.25ms</td> <td nowrap> 4.16ms</td> <td nowrap> :green_circle: -2.19%</td> <td nowrap> 6.12ms</td> <td nowrap> 6.17ms</td> <td nowrap> :white_circle: 0.90%</td> <td nowrap> :green_circle: 49%</td></tr><tr><td><tt>execute/</tt><br><tt>fibonacci_tail</td> <td nowrap> 892.56µs</td> <td nowrap> 917.01µs</td> <td nowrap> :red_circle: 2.99%</td> <td nowrap> 1.53ms</td> <td nowrap> 1.53ms</td> <td nowrap> :white_circle: 0.00%</td> <td nowrap> :yellow_circle: 67%</td></tr><tr><td><tt>execute/</tt><br><tt>global_bump</td> <td nowrap> 801.15µs</td> <td nowrap> 799.03µs</td> <td nowrap> :white_circle: -0.34%</td> <td nowrap> 1.72ms</td> <td nowrap> 1.72ms</td> <td nowrap> :white_circle: -0.02%</td> <td nowrap> :red_circle: 115%</td></tr><tr><td><tt>execute/</tt><br><tt>global_const</td> <td nowrap> 775.13µs</td> <td nowrap> 739.71µs</td> <td nowrap> :green_circle: -3.94%</td> <td nowrap> 1.37ms</td> <td nowrap> 1.37ms</td> <td nowrap> :white_circle: -0.10%</td> <td nowrap> :yellow_circle: 85%</td></tr><tr><td><tt>execute/</tt><br><tt>host_calls</td> <td nowrap> 36.59µs</td> <td nowrap> 36.31µs</td> <td nowrap> :white_circle: -0.68%</td> <td nowrap> 39.24µs</td> <td nowrap> 39.27µs</td> <td nowrap> :white_circle: -0.08%</td> <td nowrap> :green_circle: 8%</td></tr><tr><td><tt>execute/</tt><br><tt>memory_fill</td> <td nowrap> 1.21ms</td> <td nowrap> 1.22ms</td> <td nowrap> :white_circle: 0.06%</td> <td nowrap> 2.23ms</td> <td nowrap> 2.23ms</td> <td nowrap> :white_circle: -0.05%</td> <td nowrap> :yellow_circle: 84%</td></tr><tr><td><tt>execute/</tt><br><tt>memory_sum</td> <td nowrap> 1.15ms</td> <td nowrap> 1.15ms</td> <td nowrap> :white_circle: 0.04%</td> <td nowrap> 2.26ms</td> <td nowrap> 2.26ms</td> <td nowrap> :white_circle: 0.04%</td> <td nowrap> :yellow_circle: 95%</td></tr><tr><td><tt>execute/</tt><br><tt>memory_vec_add</td> <td nowrap> 2.38ms</td> <td nowrap> 2.37ms</td> <td nowrap> :white_circle: -0.11%</td> <td nowrap> 4.43ms</td> <td nowrap> 4.41ms</td> <td nowrap> :white_circle: -1.07%</td> <td nowrap> :yellow_circle: 86%</td></tr><tr><td><tt>execute/</tt><br><tt>recursive_is_even</td> <td nowrap> 690.95µs</td> <td nowrap> 693.06µs</td> <td nowrap> :white_circle: 0.34%</td> <td nowrap> 993.76µs</td> <td nowrap> 989.28µs</td> <td nowrap> :white_circle: -0.16%</td> <td nowrap> :green_circle: 43%</td></tr><tr><td><tt>execute/</tt><br><tt>recursive_ok</td> <td nowrap> 99.43µs</td> <td nowrap> 99.60µs</td> <td nowrap> :white_circle: -0.25%</td> <td nowrap> 145.40µs</td> <td nowrap> 144.85µs</td> <td nowrap> :white_circle: -0.52%</td> <td nowrap> :green_circle: 45%</td></tr><tr><td><tt>execute/</tt><br><tt>recursive_scan</td> <td nowrap> 134.98µs</td> <td nowrap> 134.68µs</td> <td nowrap> :white_circle: -0.14%</td> <td nowrap> 190.12µs</td> <td nowrap> 191.58µs</td> <td nowrap> :white_circle: 0.78%</td> <td nowrap> :green_circle: 42%</td></tr><tr><td><tt>execute/</tt><br><tt>recursive_trap</td> <td nowrap> 9.60µs</td> <td nowrap> 9.90µs</td> <td nowrap> :red_circle: 3.04%</td> <td nowrap> 14.27µs</td> <td nowrap> 14.24µs</td> <td nowrap> :white_circle: -0.14%</td> <td nowrap> :green_circle: 44%</td></tr><tr><td><tt>execute/</tt><br><tt>regex_redux</td> <td nowrap> 458.75µs</td> <td nowrap> 459.85µs</td> <td nowrap> :white_circle: 0.22%</td> <td nowrap> 796.70µs</td> <td nowrap> 803.27µs</td> <td nowrap> :white_circle: 0.84%</td> <td nowrap> :yellow_circle: 75%</td></tr><tr><td><tt>execute/</tt><br><tt>rev_complement</td> <td nowrap> 434.18µs</td> <td nowrap> 434.56µs</td> <td nowrap> :white_circle: 0.08%</td> <td nowrap> 801.27µs</td> <td nowrap> 799.97µs</td> <td nowrap> :white_circle: -0.09%</td> <td nowrap> :yellow_circle: 84%</td></tr><tr><td><tt>execute/</tt><br><tt>tiny_keccak</td> <td nowrap> 325.92µs</td> <td nowrap> 323.77µs</td> <td nowrap> :white_circle: -0.67%</td> <td nowrap> 678.17µs</td> <td nowrap> 671.21µs</td> <td nowrap> :white_circle: -0.69%</td> <td nowrap> :red_circle: 107%</td></tr><tr><td><tt>execute/</tt><br><tt>trunc_f2i</td> <td nowrap> 740.09µs</td> <td nowrap> 741.37µs</td> <td nowrap> :white_circle: 0.14%</td> <td nowrap> 1.56ms</td> <td nowrap> 1.56ms</td> <td nowrap> :white_circle: 0.00%</td> <td nowrap> :red_circle: 111%</td></tr><tr><td><tt>instantiate/</tt><br><tt>wasm_kernel</td> <td nowrap> 53.10µs</td> <td nowrap> 54.95µs</td> <td nowrap> :red_circle: 4.25%</td> <td nowrap> 53.74µs</td> <td nowrap> 57.76µs</td> <td nowrap> :red_circle: 6.02%</td> <td nowrap> :green_circle: 5%</td></tr><tr><td><tt>translate/</tt><br><tt>erc1155</td> <td nowrap> 195.60µs</td> <td nowrap> 197.00µs</td> <td nowrap> :white_circle: 0.83%</td> <td nowrap> 435.65µs</td> <td nowrap> 438.24µs</td> <td nowrap> :white_circle: 0.57%</td> <td nowrap> :red_circle: 122%</td></tr><tr><td><tt>translate/</tt><br><tt>erc20</td> <td nowrap> 97.72µs</td> <td nowrap> 97.45µs</td> <td nowrap> :white_circle: -0.43%</td> <td nowrap> 207.46µs</td> <td nowrap> 209.15µs</td> <td nowrap> :white_circle: 0.77%</td> <td nowrap> :red_circle: 115%</td></tr><tr><td><tt>translate/</tt><br><tt>erc721</td> <td nowrap> 137.49µs</td> <td nowrap> 138.46µs</td> <td nowrap> :white_circle: 0.82%</td> <td nowrap> 300.91µs</td> <td nowrap> 304.44µs</td> <td nowrap> :white_circle: 1.06%</td> <td nowrap> :red_circle: 120%</td></tr><tr><td><tt>translate/</tt><br><tt>spidermonkey</td> <td nowrap> 61.34ms</td> <td nowrap> 61.45ms</td> <td nowrap> :white_circle: 0.17%</td> <td nowrap> 0.00ns</td> <td nowrap> 0.00ns</td> <td nowrap> :white_circle: 0.63%</td> <td nowrap> :green_circle: -100%</td></tr><tr><td><tt>translate/</tt><br><tt>wasm_kernel</td> <td nowrap> 3.94ms</td> <td nowrap> 3.96ms</td> <td nowrap> :white_circle: 0.51%</td> <td nowrap> 7.84ms</td> <td nowrap> 7.92ms</td> <td nowrap> :white_circle: 1.12%</td> <td nowrap> :yellow_circle: 100%</td></tr> </tbody> </table> 

 [Link to pipeline](https://gitlab.parity.io/parity/mirrors/wasmi/-/jobs/2912648)

File: crates/core/src/trap.rs [hunk 1]
```diff
@@ -216,7 +216,7 @@ impl StdError for Trap {
 /// See [`Trap`] for details.
 ///
 /// [`Trap`]: struct.Trap.html
-#[derive(Debug, Copy, Clone)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
```
Comment (yjhmelody) [inline crates/core/src/trap.rs:219]: Could you cherry pick this line to master firstly?

Comment (Robbepop) [reply crates/core/src/trap.rs:219]: No, this PR is WIP for good reason. Many, if not everything here is subject to change. We can talk about this once the PR is no longer considered WIP.

File: crates/wasmi/src/engine/bytecode2/mod.rs [hunk 2]
```diff
@@ -0,0 +1,3895 @@
+#![allow(dead_code)] // TODO: remove
```
Comment (yjhmelody) [inline crates/wasmi/src/engine/bytecode2/mod.rs:1]: Why not rename `bytecode2` to `regmach`? The tests part use `regmach` name.

Comment (Robbepop) [reply crates/wasmi/src/engine/bytecode2/mod.rs:1]: This is just a WIP state. The final goal is to entirely replace the old stack machine based `wasmi` executor and bytecode.

Comment (yjhmelody) [reply crates/wasmi/src/engine/bytecode2/mod.rs:1]: You means only exist one kind instruction set ?
Not good news for me...

Comment (Robbepop) [reply crates/wasmi/src/engine/bytecode2/mod.rs:1]: Yes, we do not want to support and maintain multiple `wasmi` bytecodes and execution engines.

> Not good news for me...

Why? Can you please elaborate?

Comment (Robbepop) [reply crates/wasmi/src/engine/bytecode2/mod.rs:1]: @yjhmelody I would really appreciate some elaboration.

**Comment by Robbepop** [comment]
Note: The Wasm benchmark CI is currently failing because of this `rustc` bug: https://github.com/rust-lang/rust/issues/114725

**Comment by Robbepop** [comment]
Note: https://github.com/paritytech/wasmi/pull/729/commits/af9040f6686d3f45983af78147e08f6b1f1a7580 fixes the benchmarking CI due to this `rustc` or LLVM bug: https://github.com/rust-lang/rust/issues/114725

**Comment by Robbepop** [comment]
As discussed with @athei I will merge this PR now and start working on the remaining TODO items in isolation via follow-up PRs. For this process I am going to write a bunch of issues to track their progress.

**Pull request merged**